### PR TITLE
Fix building bibs not blocking placement.

### LINKED ISSUE #18816: Bot places buildings on bibs and exits blocking itself
image

image

### LINKED ISSUE #18840: Structures can be placed on refinery pads
## Issue Summary

DESCRIPTION
In the latest dev build (on 19 Nov) is now possible to place structures on refinery pads, also blocking the harvesters.
refpad-block
refpad-block-2

This was not possible in the latest release.

A git bisect revealed that daa8c74c375d88999fa4671cecb45fe4222d2548 is the breaking commit; it makes buildings placeable
in shadows, but also makes them placeable on refinery pads.

## Reproduction

STEPS TO REPRODUCE THE ISSUE
1. Build a refinery
2. Build something else and place it on the refinery pad
3. The harvester is blocked as well

## Expected behavior

EXPECTED BEHAVIOR
It should not be placeable

## Screenshots / Screen recordings / Replays

refpad-block
refpad-block-2

### PR DESCRIPTION
**Pull request by pchote**
Fixes #18816, which was regressed by #18688 removing the `BuildingInfluence` check from `BuildingUtils.IsCellBuildable`. 
Fixes #18840.

Bib cells are not added to the `ActorMap`, so were no longer being considered for blocking building placement. This PR solves the regression by adding a new set of methods to explicitly track cells that are blocked for building.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/World/ActorMap.cs [hunk 1]
```diff
@@ -616,6 +618,39 @@ public IEnumerable<Actor> ActorsInBox(WPos a, WPos b)
...
+		public void RemoveBuildingBlocker(object blocker)
+		{
+			var footprint = buildingBlockers[blocker];
+			buildingBlockers.Remove(blocker);
+
+			foreach (var c in footprint)
+			{
+				var blockers = buildingBlocked[c];
+				blockers.Remove(blocker);
+				if (blockers.Count == 0)
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/World/ActorMap.cs:650]: This is a bit weird. You are only removing it for real if nothing is left so that the removal is complete (all or nothing)? Why is this necessary?

Comment (pchote) [reply OpenRA.Mods.Common/Traits/World/ActorMap.cs:650]: We are taking advantage of the fact that the total number of blocked cells on the map is small and sparse to use a dictionary of (blocked cell) &rarr; (list of blockers) instead of having to define an entire `CellLayer`. We can then simply check if the dictionary contains the cell key - if any values are defined then the cell is blocked, and we don't care about *what* is blocking.

When we remove an item we do care about what is blocking, and if this was the last blocker we need to remove the empty list so that the `ContainsKey` check can correctly return false.

File: OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs [hunk 2]
```diff
@@ -70,7 +70,8 @@ public static bool IsCellBuildable(this World world, CPos cell, ActorInfo ai, Bu
 			}
 
 			// Buildings can never be placed on ramps
-			return world.Map.Ramp[cell] == 0 && bi.TerrainTypes.Contains(world.Map.GetTerrainInfo(cell).Type);
+			return world.Map.Ramp[cell] == 0 && bi.TerrainTypes.Contains(world.Map.GetTerrainInfo(cell).Type) &&
+				!world.ActorMap.IsBuildingBlocked(cell);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:74]: I suggest moving those checks up to a) prevent `IsBuildingBlocked` from having an effect on `AllowInvalidPlacement` and b) save us from searching actors in a cell that is not buildable anyway.
```csharp
			if (!world.Map.Contains(cell))
				return false;

			// Buildings can never be placed on ramps
			if (world.Map.Ramp[cell] != 0 || !bi.TerrainTypes.Contains(world.Map.GetTerrainInfo(cell).Type))
				return false;

			if (bi.AllowInvalidPlacement)
				return true;

			if (world.ActorMap.IsBuildingBlocked(cell))
				return false;

			// Replaceable actors are rare, so avoid initializing state unless we have to
			var checkReplacements = ai != null && ai.HasTraitInfo<ReplacementInfo>();
			HashSet<string> acceptedReplacements = null;

			var foundActors = false;
			foreach (var a in world.ActorMap.GetActorsAt(cell))
			{
				if (a == toIgnore)
					continue;

				// If this is potentially a replacement actor we must check *all* cell occupants
				// before we know the placement is invalid
				// Otherwise, we can bail immediately
				if (!checkReplacements)
					return false;

				foundActors = true;
				foreach (var r in a.TraitsImplementing<Replaceable>())
				{
					if (r.IsTraitDisabled)
						continue;

					if (acceptedReplacements == null)
						acceptedReplacements = new HashSet<string>();

					acceptedReplacements.UnionWith(r.Info.Types);
				}
			}

			// Replacements are enabled and the cell contained at least one (not ignored) actor
			if (foundActors)
			{
				// The cell contains at least one actor, and none were replaceable
				if (acceptedReplacements == null)
					return false;

				// The cell contains at least one replaceable actor, but not of the types we accept
				var foundReplacement = ai.TraitInfos<ReplacementInfo>()
					.Any(r => r.ReplaceableTypes.Overlaps(acceptedReplacements));

				if (!foundReplacement)
					return false;
			}

			return true;
```

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:74]: On a deeper look, I don't think this is correct. `IsBuildingBlocked` is supposed to have an effect on `AllowInvalidPlacement`: the expected behaviour is that those cells should show red on the placement preview; the overall placement is allowed, but those specific tiles are ignored.

The suggested changes regress the concrete placement behaviour in d2k: hovering the concrete PBOG over a building that that was built on concrete correctly shows red, but hovering the concrete PBOG over a building that was built on rock shows white!

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:74]: In the longer term (i'm not sure its feasible for the upcoming playtest) it would probably be better to use the yellow placement tiles instead of red in this case - red is supposed to mean that the placement is not allowed at all, whereas yellow means it is allowed but those specific cells are going to behave differently.

If we can get this and #18685 merged in a reasonable time we can look into the feasibility of removing `AllowInvalidPlacement` from building and instead introducing a new `D2kConcrete` trait that can be placed using PBOG but behaves like `AllowInvalidPlacement` and has the special case handling for yellow tiles in `D2kActorPreviewPlaceBuildingPreview`.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:74]: > The suggested changes regress the concrete placement behaviour in d2k: hovering the concrete PBOG over a building that that was built on concrete correctly shows red, but hovering the concrete PBOG over a building that was built on rock shows white!

I actually expected something like this, and think it would be better to always show white not red, simply because a full red preview won't block you from placing it down (when you would expect it to reject placement; I think we have an open issue for that). I guess this doesn't matter if we remove `AllowInvalidPlacement` in the long run.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:74]: My idea of a specific `D2kConcrete` and PBOG/preview handling could fix that too.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:74]: Btw, I just remembered what is going on here. Whatever we do in this method is kind of a moot point because of https://github.com/OpenRA/OpenRA/blob/daa8c74c375d88999fa4671cecb45fe4222d2548/OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs#L76-L79
Which means returning false here just results in a red color that is completely ignored.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Buildings/BuildingUtils.cs:74]: Indeed. That's what I said above 😄

**Review by abcdefg30 — APPROVED** [review]
Ok.

**Comment by pchote** [comment]
Considering #18852, we should instead revert #18689 and then change `BuildingUtils.IsCellBuildable` to query the BuildingMap again (there is no need to throw out the other improvements from #18688).

I have filed #18853 outling what I believe to be the better long term solution, which is too big to do in this release cycle.
